diff --git a/src/utils.cpp b/src/utils.cpp
index 4a0e53d..d7b020d 100644
--- a/src/utils.cpp
+++ b/src/utils.cpp
@@ -27,6 +27,10 @@

 #include <linux/version.h>

+#ifdef __ANDROID__ // Use ghc filesystem(https://github.com/gulrak/filesystem)
+#include <ghc/filesystem.hpp>
+namespace std_filesystem = ghc::filesystem;
+#else
 #if __has_include(<filesystem>)
 #include <filesystem>
 namespace std_filesystem = std::filesystem;
@@ -36,6 +40,7 @@ namespace std_filesystem = std::experimental::filesystem;
 #else
 #error "neither <filesystem> nor <experimental/filesystem> are present"
 #endif
+#endif

 namespace {

@@ -339,7 +344,11 @@ namespace {

   std::string unpack_kheaders_tar_xz(const struct utsname& utsname)
   {
+#ifdef __ANDROID__
+    std::string path_prefix{"/data/local/tmp"};
+#else
     std::string path_prefix{"/tmp"};
+#endif
     if (const char* tmpdir = ::getenv("TMPDIR")) {
       path_prefix = tmpdir;
     }
